[JS/Web]Added FusedConv.

satyajandhyala opened the pull request:
### Description
Added FusedConv and FusedConvTranspose



### Motivation and Context
<!-- - Why is this change required? What problem does it solve?
- If it fixes an open issue, please link to the issue here. -->
Improve performance

snnn commented:
/azp run Windows CPU CI Pipeline

satyajandhyala commented:
/azp run WIndows GPU CI Pipeline

edgchen1 commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -263,6 +275,10 @@ void RegisterConvActivationFusionRules(SelectorActionRegistry& registry) {
                                      std::move(selector), std::move(action));
+  auto selector2 = std::make_unique<selectors::ConvActivationSelector>();
+  auto action2 = std::make_unique<actions::FuseConvActivationAction>(kMSInternalNHWCDomain);
+  registry.RegisterSelectorAndAction("MSInternalNHWCConvAct", {{"Conv", {1, 11}}},
```
we want to register the same names across minimal and full builds. otherwise, the minimal build would be unable to apply the action associated with that name.

the pattern should be:
```c++
#if !defined(ORT_MINIMAL_BUILD)
  registry.RegisterSelectorAndAction(name, ...);
#else
  registry.RegisterAction(name, ...);
#endif
```

edgchen1 commented on onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.h:
```diff
@@ -81,21 +83,22 @@ class SelectorActionRegistry {
                                  std::unique_ptr<NodeSelector> selector,
-                                 std::unique_ptr<Action> action);
+                                 std::unique_ptr<Action> action,
+                                 const std::string& domain = kOnnxDomain);
```
I'd think that the domain should be part of the specification of the ops that a selector/action entry applies to (until now this specification was `ops_and_versions`). then we could have a single selector/action entry that can handle ops from multiple domains.

satyajandhyala replied:
The Action object is tied to a specific domain due to the API
https://github.com/microsoft/onnxruntime/blob/ad511bc757e50d16d043d8d45056a36c8140f848/onnxruntime/core/optimizer/selectors_actions/actions.h#L133

satyajandhyala replied:
Each Action object is tied to a specific Domain.

skottmckay replied:
Not quite following. The RuntimeState is provided as an argument, and the Domain method is overridable. Why can't the implementation of a specific Action that needs to be dynamic based on the RuntimeState's target node do that?

satyajandhyala replied:
I guess you are asking why cant we use one Action and keep changing the domain as we work through different items?
I think due to the class hierarchy and base class implementing some functionality, it is error prone.  Making that change requires modifying base class(es). I think it is cleaner creating another object.

skottmckay replied:
Ed and I discussed this further (as we've done the most work on the select/action setup) and the consensus was to not push the domain check up to this level unless we're going to switch to using OpIdentifier for the checks. Doing so limits the functionality by restricting the transformer to a single domain which is undesirable. 

For now it should be fine to add the internal NHWC domain to the initial check (domain is kOnnxDomain or kMSDomain) and use the existing code.

The selector can add any logic required around ops in the internal NHWC domain. That should be minimal as there are limited EPs that request NHWC and the only other one using L2 optimizers is the CUDA EP. From a look at the L2 optimizers only the ones involving Conv could potentially have a node in the NHWC domain. 

SelectorActionTransformer is only used for L2 and above optimizers as its intended for usage in a minimal build where L1 happens during creation of the ORT format model.

satyajandhyala replied:
The domain (if it is not kOnnxDomain) now specified as a part of the operator in the OpsVersionsMap.

edgchen1 commented on onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.cc:
```diff
@@ -32,33 +33,34 @@ void SelectorActionRegistry::RegisterSelectorAndAction(const std::string& name,
   for (const auto& [op_type, versions] : entry.ops_and_versions) {
     ORT_UNUSED_PARAMETER(versions);
-    op_type_to_entry_.emplace(op_type, &entry);
+    op_type_to_entry_.emplace(op_type + domain, &entry);
```
if we go with this approach, would be good to use something like `pair<string, string>` or at least separate them with a delimiter. if the op type and domain can be arbitrary strings the concatenation of one pair could be the same as another distinct pair.

satyajandhyala replied:
I thought about picking pair<string, string> as the key. However I thought it might add the complexity of the hash table.

satyajandhyala replied:
In this case we are not uniquely identifying the op, (Domain, op, version) in the Key. The key is (domain, operator) the version is in Entry which maps different versions.

skottmckay replied:
I assumed we wouldn't use a map of op to version if we switched to using OpIdentifier. 

Given the primary usage is in MatchAndProcess by this code, if we use OpIdentifier we can just do a direct lookup instead of lookup of op followed by matching of version. 

```
      // check the supported versions if specified
      const auto& versions = entry->ops_and_versions.find(node.OpType())->second;
      if (!versions.empty()) {
        if (std::find(versions.cbegin(), versions.cend(), node.SinceVersion()) == versions.cend()) {
          continue;
        }
      }
```

OpIdentifier may need the ability to do a wildcarded match of the version but that could probably be implemented in an operator== method to keep things simple.

satyajandhyala replied:
I was not planning on changing the existing functionality. My aim is to make this work for domains other than the default domain, kOnnxDomain.

skottmckay replied:
If we want to minimize changes, the selector could apply any filtering on domain instead of pushing that up to the transformer level. i.e. the filtering on op and opset in MatchAndProcess can be seen as a simplistic check, with the implementation of the selector doing additional things like domain/EP validation. With that a lot of the changes in the PR may not be necessary. 

This might also be better given potentially the EP involved needs to be checked as well before deciding in the fusion can proceed. 

e.g. the CUDA EP how has some NHWC ops, and the ConvActivationFusion EP list in GenerateTransformers includes that EP. do you know if the CUDA EP expects a Conv in the internal NHWC domain to be fused?

skottmckay commented on onnxruntime/core/providers/js/operators/conv.h:
```diff
@@ -9,14 +9,27 @@
 namespace js {
 
-template <bool is_channels_last>
+template <bool is_channels_last, bool is_fused_conv = false>
```
nit: this approach of using a template for what could be a boolean arg to the ctor of the implementation is extremely inefficient in terms of binary size as the entire implementation is copied for each permutation of the template params. 

instead you can create lightweight wrapper classes to plugin the values.

e.g. change the base class ctor to `Conv(const OpKernelInfo& info, bool is_channels_last, bool is_fused_conv)` and add a templatized wrapper class to plugin the booleans. That way only the code of the wrapper class is templatized and there's a single copy of the implementation.

```c++
template <bool is_channels_last, bool is_fused_conv = false>
class ConvUsage : public Conv {
 public:
  ConvUsage(const OpKernelInfo& info) 
    : Conv(info, is_channels_last, is_fused_conv) {
};
```

skottmckay commented on onnxruntime/core/optimizer/graph_transformer_utils.cc:
```diff
@@ -294,7 +295,7 @@ InlinedVector<std::unique_ptr<GraphTransformer>> GenerateTransformers(
       transformers.emplace_back(std::make_unique<DynamicQuantizeMatMulFusion>(cpu_ep));
 
-      transformers.emplace_back(std::make_unique<ConvActivationFusion>(cpu_cuda_rocm_acl_armnn_eps));
+      transformers.emplace_back(std::make_unique<ConvActivationFusion>(cpu_cuda_rocm_acl_armnn_js_eps));
```
Does this do the fusion with ConvTranspose or is another optimizer and new unit tests required?

satyajandhyala replied:
Currently ConvActivationFusion only works on Conv operator. For fusing ConvTranspose nodes with activation we need to add a new transformation, ConvTransposeActivationFusion (not included in this PR). The JSEP code works if the input graph has already has ConvTranspose and actions fused.

satyajandhyala replied:
Currently this is only for Conv. For ConvTranspose I just need to register selection-actions pair using the same Selector and Actor used for Conv only using opcode "CovTranspose" instead of "Conv".

satyajandhyala replied:
Adding ConvTranspose also.

satyajandhyala replied:
Backed out ConvTranspose

edgchen1 replied:
> Backed out ConvTranspose

is it still here?

https://github.com/microsoft/onnxruntime/blob/1834ab8512128c06626e70a09a4fb441912fa4db/onnxruntime/core/optimizer/conv_activation_fusion.cc#L270

skottmckay commented on onnxruntime/core/graph/contrib_ops/contrib_defs.cc:
```diff
@@ -1593,6 +1594,84 @@ activation.)DOC")
                                   ONNX_NAMESPACE::convPoolShapeInference(ctx, true, false, 0, 1);
                                 }));
 
+ONNX_MS_OPERATOR_SET_SCHEMA(FusedConvTranspose, 1,
```
Please add C++ unit tests so the schema/shape inferencing/etc. is validated in a basic build (i.e. CPU EP only).

satyajandhyala replied:
Is there an example test for this ?

satyajandhyala replied:
Not applicable anymore as FusedConvTranspose change is reverted.

skottmckay commented on onnxruntime/contrib_ops/js/js_contrib_kernels.cc:
```diff
@@ -11,6 +11,8 @@ class ONNX_OPERATOR_KERNEL_CLASS_NAME(kJsExecutionProvider, kMSDomain, 1, Gelu);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kJsExecutionProvider, kMSDomain, 1, BiasSplitGelu);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kJsExecutionProvider, kMSDomain, 1, BiasAdd);
 class ONNX_OPERATOR_KERNEL_CLASS_NAME(kJsExecutionProvider, kMSDomain, 1, SkipLayerNormalization);
+class ONNX_OPERATOR_KERNEL_CLASS_NAME(kJsExecutionProvider, kMSDomain, 1, FusedConv);
+class ONNX_OPERATOR_KERNEL_CLASS_NAME(kJsExecutionProvider, kMSDomain, 1, FusedConvTranspose);
```
Not quite understanding this.

When the JS EP takes a Conv or ConvTranspose node it gets converted to the internal NHWC domain during partitioning. When the level 2 optimizer fuses it (after partitioning) I expected the node to remain in the internal NHWC domain so it's clear what the layout is.

Given that, when does a FusedConv{Transpose} in the contrib domain (which would be NCHW format) come into play for the JS EP?

satyajandhyala replied:
We have a choice from the frontend to set the preference, either channel-last (NHWC) or channel-first(NCHW). If the preference is set to channel-first then we could stil apply fusion.

skottmckay commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -84,14 +84,15 @@ class ConvActivationSelector : public NodeSelector {
-          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Sigmoid", {6, 13}) ||
-          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Tanh", {6, 13}) ||
-          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "LeakyRelu", {6, 16})) {
+      const auto& domain = activation_node.Domain();
+      if (graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Relu", {6, 13, 14}, domain) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Sigmoid", {6, 13}, domain) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "Tanh", {6, 13}, domain) ||
+          graph_utils::IsSupportedOptypeVersionAndDomain(activation_node, "LeakyRelu", {6, 16}, domain)) {
```
I don't think adding domain to these checks should be necessary. We only use the internal NHWC domain for layout sensitive operators so none of the activations/Add/Clip would ever be in that domain.

skottmckay commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -275,6 +291,10 @@ void RegisterConvAddReluFusionRules(SelectorActionRegistry& registry) {
   registry.RegisterSelectorAndAction(name, {{"Conv", {1, 11}}},
                                      std::move(selector), std::move(action));
+  auto selector2 = std::make_unique<selectors::ConvAddRelu>();
+  auto action2 = std::make_unique<actions::FuseConvActivationAction>(kMSInternalNHWCDomain);
```
The ConvAddRelu Select method has this which would only make it apply to the CUDA EP wouldn't it? Is there a test that checks the new handling that uses the CUDA EP? 

```
  std::optional<NodesToOptimizeIndices> Select(const GraphViewer& graph_viewer, const Node& node) const override {
    const auto& domain = node.Domain();
    const std::string_view node_ep = node.GetExecutionProviderType();
    // only for CUDA EP
    if (node_ep != kCudaExecutionProvider) {
      return std::nullopt;
    }
```

skottmckay commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -101,19 +102,19 @@ class ConvActivationSelector : public NodeSelector {
 
     // check EP type and activation
     if (node_ep == kCudaExecutionProvider || node_ep == kRocmExecutionProvider) {
-      if (!graph_utils::IsSupportedOptypeVersionAndDomain(*next_node, "Relu", {6, 13, 14})) {
+      if (!graph_utils::IsSupportedOptypeVersionAndDomain(*next_node, "Relu", {6, 13, 14}, domain)) {
```
Should this be using the domain from the next_node and not the current node? 

e.g. if I have a Conv in the internal NHWC domain, the Relu will be in the ONNX domain and this check will reject the fusion.

satyajandhyala replied:
Dropped domain argument.

skottmckay commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -173,10 +175,15 @@ namespace actions {
 
 class FuseConvActivationAction : public ReplaceWithNew {
+ public:
+  FuseConvActivationAction(const std::string& domain = kMSDomain) : ReplaceWithNew(), domain_{domain} {}
+
  private:
-  std::string OpType(const RuntimeState&) const override { return "FusedConv"; }
+  std::string OpType(const RuntimeState&) const override {
+    return domain_ == kMSDomain ? "FusedConv" : "Conv";
+  }
 
-  std::string Domain(const RuntimeState&) const override { return kMSDomain; }
+  std::string Domain(const RuntimeState&) const override { return domain_; }
```
Can the OpType and Domain can be inferred from the target node so a domain_ member isn't required? Seem like we return kMsDomain/FusedConv unless the target node's domain is the internal NHWC domain, in which case we return kMSInternalNHWCDomain/Conv

satyajandhyala replied:
Modified code to used domain from the target node.

satyajandhyala replied:
The target domain is not set when the Domain() and Optype() methods are called on the target node.

skottmckay commented on onnxruntime/core/optimizer/conv_add_act_fusion.cc:
```diff
@@ -293,6 +295,10 @@ void RegisterConvAddActivationFusionRules(SelectorActionRegistry& registry) {
   auto selector_nhwc = std::make_unique<selectors::ConvAddActivationSelector>();
   registry.RegisterSelectorAndAction("NhwcFusedConvAct", {{"NhwcFusedConv", {1, 11}}},
                                      std::move(selector_nhwc), std::move(action_nhwc));
+  auto action_ms_internal_nhwc = std::make_unique<actions::FuseConvAddActivationAction>(kMSInternalNHWCDomain);
+  auto selector_ms_internal_nhwc = std::make_unique<selectors::ConvAddActivationSelector>();
```
I think the existing code is providing a bad example here. There's no need for ConvAddAct and NhwcFusedConvAct to be separate registrations. The action already handles both usages internally so we could combine them with 

```
  registry.RegisterSelectorAndAction("ConvAddAct", {{"Conv", {1, 11}}, {"NhwcFusedConv", {1, 11}}},
                                     std::move(selector), std::move(action));
```

And that single registration could also be used to add support for the internal NHWC domain by 
- applying any EP specific filtering in the Select when the Conv is in the internal NHWC domain, and 
- updating FuseConvAddActivationAction to check the optype and domain of the target node in OpType() and Domain(). the rules are something like this aren't they?  
  - onnx Conv -> kMsDomain FusedConv
  - internal NHWC Conv -> internal NHWC Conv
  - NhwcFusedConv -> NhwcFusedConv

satyajandhyala replied:
Yes. I figure the first part in this comment already. That is how added ConvTranspose in conv_activation_fusion.cc using single registration.

satyajandhyala commented:
I have removed FusedConvTranspose because adding the new schema requires a C/C++ unit test to prioritize adding only FusedConv which is the primary requirement.
I have incorporated all the feedback I received as of yesterday.

skottmckay commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -260,9 +260,16 @@ void RegisterConvActivationFusionRules(SelectorActionRegistry& registry) {
   const auto name = "ConvAct";
   auto action = std::make_unique<actions::FuseConvActivationAction>();
 #if !defined(ORT_MINIMAL_BUILD)
+  SelectorActionRegistry::OpVersionsMap op_versions_map;
```
Does FuseConvActivationAction need to be updated to return kMSInternalNHWCDomain if the the target node is in that domain?

satyajandhyala replied:
The implementation of the class needs to be modified if that is what you are asking; currently it only work for Conv op in ONNX domain (target node op FusedConv in MsDomain). FuseConvActivationAction needs to supply the target domain. It does not come from anywhere else. The object itself doesn't need to be updated. However when we create this object we create it for specific domain otherwise no why to make it work different domains. Conv op node from kOnnxDomain will be translated to FusedConv in kMsDomain as you already know. NhwcConv op from kMsInternalNhwcDomain goes to FusedNhwcConv in kMsInternalNhwcDomain.

skottmckay commented on onnxruntime/core/optimizer/conv_add_act_fusion.cc:
```diff
@@ -207,14 +209,34 @@ using NTO = NodesToOptimize;
  private:
-  std::string OpType(const RuntimeState& runtimeState) const override {
-    return (runtimeState.selected_nodes.Target().OpType() == "Conv") ? "FusedConv" : "NhwcFusedConv";
+  std::string OpType(const RuntimeState& runtime_state) const override {
+    const auto& domain = runtime_state.selected_nodes.Target().Domain();
+    const auto& op_type = runtime_state.selected_nodes.Target().OpType();
+    if (domain == kMSDomain) {
+      if (op_type == "Conv") {
+        return "FusedConv";
+      } else if (op_type == "NhwcConv") {
+        return "NhwcFusedConv";
+      } else {
+        ORT_THROW("Unsupported op type: ", op_type);
+      }
+    } else if (domain == kMSInternalNHWCDomain) {
+      if (op_type != "Conv") {
+        ORT_THROW("Unsupported op type: ", op_type);
+      }
+      return "Conv";
+    } else {
+      ORT_THROW("Unsupported domain: ", domain);
+    }
```
Won't the target node be in the ONNX domain for fusing ONNX Conv into kMSDomain:FusedConv?

satyajandhyala replied:
That is right. Based on the previous comments I thought that the domain will be already set to kMSDOmain on the target node.

satyajandhyala commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -263,6 +275,10 @@ void RegisterConvActivationFusionRules(SelectorActionRegistry& registry) {
   registry.RegisterSelectorAndAction(name, {{"Conv", {1, 11}}},
                                      std::move(selector), std::move(action));
+  auto selector2 = std::make_unique<selectors::ConvActivationSelector>();
+  auto action2 = std::make_unique<actions::FuseConvActivationAction>(kMSInternalNHWCDomain);
```
Modified code to use single Selector-Action pair.

satyajandhyala commented:
@skottmckay It seems the domain is not set on the target node in runtime_state. Resulting in many test failures.
The graph transformation is expected to provide it. Looking at the code here https://github.com/microsoft/onnxruntime/blob/8ebdd3bbcafc7e195e41b2aa3881d258d365f608/onnxruntime/core/optimizer/selectors_actions/actions.cc#L104C6-L104C6
If that is the case the graph transformer needs to know the target domain and op_type. I think in this case based on the existing code my previous version works as it creates separate instances of the ConvAddActivation Selector and FuseConvAddActivationAction pair one for each domains, kMsDomain and kMsInternalNHWCDomain.

skottmckay commented:
> @skottmckay It seems the domain is not set on the target node in runtime_state. Resulting in many test failures. It seems Action::Domain() and Action::OpType() are expected to provide target domain and op_type.

I believe the domain is set as it's required to do schema lookup in Graph::Resolve which happens before any optimizations take place. The ONNX domain is an empty string.

skottmckay commented on onnxruntime/core/optimizer/conv_add_act_fusion.cc:
```diff
@@ -207,14 +209,34 @@ using NTO = NodesToOptimize;
+      }
+      return "Conv";
+    } else {
+      ORT_THROW("Unsupported domain: ", domain);
```
nit: multiple ORT_THROW statements increase the binary size vs. a single one at the end.

edgchen1 commented on onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.h:
```diff
@@ -86,7 +95,7 @@ class SelectorActionRegistry {
 
   // register an action
-  void RegisterAction(const std::string& name, std::unique_ptr<Action> action);
+  void RegisterAction(const std::string& name, std::unique_ptr<Action> action, const std::string& domain = kOnnxDomain);
```
do we still need a domain parameter here?

satyajandhyala replied:
Removed the domain parameter.

edgchen1 commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -260,9 +260,16 @@ void RegisterConvActivationFusionRules(SelectorActionRegistry& registry) {
 #if !defined(ORT_MINIMAL_BUILD)
+  SelectorActionRegistry::OpVersionsMap op_versions_map;
+  std::string kMSInternalNHWCDomainConv = std::string(kMSInternalNHWCDomain) + ":Conv";
+  std::string kMSInternalNHWCDomainConvTranspose = std::string(kMSInternalNHWCDomain) + ":ConvTranspose";
```
nit: the 'k' prefix implies const per the [style guide](https://google.github.io/styleguide/cppguide.html#Constant_Names) so these should be const

satyajandhyala replied:
removed the first letter k.

edgchen1 commented on onnxruntime/core/optimizer/conv_add_act_fusion.cc:
```diff
@@ -281,18 +303,23 @@ class FuseConvAddActivationAction : public ReplaceWithNew {
       };
     }
   }
+  std::string domain_;
```
is this still being used?

satyajandhyala replied:
Removed

edgchen1 commented on onnxruntime/core/optimizer/conv_add_act_fusion.cc:
```diff
@@ -69,7 +71,7 @@ class ConvAddActivationSelector : public NodeSelector {
     // 3 find the next node and check if it's a activation node, if yes, we will fuse conv+add+activation or conv+add
     //
-    if (graph_utils::IsSupportedOptypeVersionAndDomain(*add_node, "Add", {7, 13, 14})) {
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*add_node, "Add", {7, 13, 14}, node.Domain())) {
```
we're only looking for ONNX Add right? do we need to pass in `node.Domain()` here?

satyajandhyala replied:
Removed node.Domain()

satyajandhyala commented:
> > @skottmckay It seems the domain is not set on the target node in runtime_state. Resulting in many test failures. It seems Action::Domain() and Action::OpType() are expected to provide target domain and op_type.
> 
> I believe the domain is set as it's required to do schema lookup in Graph::Resolve which happens before any optimizations take place. The ONNX domain is an empty string.

I found errors even after optimizing if the domain and operator is not consistent, for example if the domain is empty string for kMsOnnxDomain and the op_type is FusedConv then it will cause error.

satyajandhyala commented:
> > @skottmckay It seems the domain is not set on the target node in runtime_state. Resulting in many test failures. It seems Action::Domain() and Action::OpType() are expected to provide target domain and op_type.
> 
> I believe the domain is set as it's required to do schema lookup in Graph::Resolve which happens before any optimizations take place. The ONNX domain is an empty string.
Consistent domain and op_type is required even after optimizing not just to do schema look up in Grpah:Resolve. It seems the consistency is check even during/after the optimization. I will have to restore my previous approach, creating different Selectior-Action pairs for different domains.

skottmckay commented on onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.cc:
```diff
@@ -39,10 +39,10 @@ void SelectorActionRegistry::RegisterSelectorAndAction(const std::string& name,
 
 void SelectorActionRegistry::RegisterAction(const std::string& name,
-                                            std::unique_ptr<Action> action) {
+                                            std::unique_ptr<Action> actionn) {
```
```suggestion
                                            std::unique_ptr<Action> action) {
```

satyajandhyala replied:
fixed typo

skottmckay commented on onnxruntime/core/optimizer/conv_add_act_fusion.cc:
```diff
@@ -287,12 +281,16 @@ class FuseConvAddActivationAction : public ReplaceWithNew {
   auto action = std::make_unique<actions::FuseConvAddActivationAction>();
   auto selector = std::make_unique<selectors::ConvAddActivationSelector>();
-  registry.RegisterSelectorAndAction("ConvAddAct", {{"Conv", {1, 11}}},
+  SelectorActionRegistry::OpVersionsMap op_versions_map;
+  const std::string kMSDomainFusedConv = std::string(kMSDomain) + ":FusedConv";
+  const std::string kMSInternalNHWCDomainNhwcFusedConv = std::string(kMSInternalNHWCDomain) + ":NhwcFusedConv";
```
If this is not being used for the JS EP yet and there are no unit tests we should exclude the changes from this PR. 

e.g. I believe NhwcFusedConv is in the contrib ops kMSDomain so using kMSInternalNHWCDomain may not match anything and you won't get the expected optimization. we should have unit tests to cover this to validate it works correctly.

satyajandhyala replied:
Reverted changes to conv_add_act_fusion.cc

skottmckay commented:
Can you please address the lint warnings around line length and missing #includes?

satyajandhyala commented on js/web/lib/wasm/jsep/webgpu/ops/3rd-party/activation_util.ts:
```diff
@@ -41,12 +41,11 @@ export const activationFnSnippet =
 export const biasActivationSnippet = (hasBias: boolean, activation?: Activation): string => `
       ${hasBias ? 'value = value + getBiasByOutputCoords(coords);' : ''}
-      ${activation ? 'value = activation(value, coords);' : ''}
+      // ${activation ? 'value = activation(value, coords);' : ''}
```
Added to a TODO to use the code when activation functionality is implemented. Leaving it here to make it easy to know where to insert it.

satyajandhyala commented:
> Can you please address the lint warnings around line length and missing #includes?

I don't see any lint warning about missing include files. Wouldn't it be an error if a required header file is not included?

edgchen1 commented:
> > Can you please address the lint warnings around line length and missing #includes?
> 
> I don't see any lint warning about missing include files. Wouldn't it be an error if a required header file is not included?

see the warnings here:
https://github.com/microsoft/onnxruntime/pull/17766/checks?check_run_id=18234199008

satyajandhyala commented:
> > > Can you please address the lint warnings around line length and missing #includes?
> > 
> > 
> > I don't see any lint warning about missing include files. Wouldn't it be an error if a required header file is not included?
> 
> see the warnings here: https://github.com/microsoft/onnxruntime/pull/17766/checks?check_run_id=18234199008

Yeah. I saw the warnings.

edgchen1 commented on onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.cc:
```diff
@@ -93,20 +93,15 @@ static Status MatchAndProcess(
-    const auto selector_action_entries = selector_action_registry.LookUpByOpType(node.OpType());
+    const auto selector_action_entries = selector_action_registry.LookUpByOpTypeAndDomain(node.OpType(), node.Domain());
+    const auto& domain = node.Domain();
+    std::string domain_optype = (domain == kOnnxDomain) ? node.OpType() : (domain + ":" + node.OpType());
```
maybe a helper function would be useful. e.g.,
```c++
std::string OpVersionsMapKey(std::string_view op_type, std::string_view domain = kOnnxDomain) {
  return (domain == kOnnxDomain)
      ? std::string{op_type}
      : std::string{domain} + ":" + std::string{op_type};
}
```
it could be used here and by callers of SelectorActionTransformer::RegisterSelectorAndAction. the callers wouldn't need to know the details of how to construct the key.

satyajandhyala replied:
Done

satyajandhyala commented on onnxruntime/core/optimizer/selectors_actions/selector_action_transformer.h:
```diff
@@ -38,6 +38,9 @@ struct NodeSelector {
  public:
+  // The key is a string representing the op, optionally specifying the domain using ':' as the
+  // separator with domain as the first part and oprator as the second part, "<domain>:<operator>".
+  // Ex: "Conv", "com.microsoft:Conv", "com.ms.internal.nhwc:Conv"
```
The domain here is the source domain. I found only conv_add_act_fusion.cc and conv_activation_fusion.cc applying transformation on nodes with op type other than ONNX domain. Other QDQ transformation apply on node with op type in ONNX domain targetting MSDomain.

satyajandhyala replied:
Clarified implied domain value,kOnnxDomain, if not specified.

satyajandhyala commented on onnxruntime/core/optimizer/conv_activation_fusion.cc:
```diff
@@ -258,18 +278,15 @@ class FuseConvAddRelu : public ReplaceWithNew {
 
 void RegisterConvActivationFusionRules(SelectorActionRegistry& registry) {
   const auto name = "ConvAct";
+  LOGS_DEFAULT(VERBOSE) << "Registering " << name << std::endl;
```
Removed the debug message

edgchen1 replied:
looks like it is still there

satyajandhyala replied:
I am still testing a change. My PR branch regressed due to some changes. I will push it as soon as done.

edgchen1 reviewed (commented):
the selector action transformer related changes look good to me. will let other folks review the rest of the PR as I'm not familiar with the JS EP.

satyajandhyala commented:
I am still test a change. I will push it as soon as done.
